bin/faucet/src/client.rs

```diff
@@ -148,137 +150,132 @@ impl FaucetClient {
-            seed,
-            block_header: root_block_header,
-            chain_mmr: root_chain_mmr,
-        }
-    }
+/// Initializes the keypair used to sign transactions.
+///
+/// If the secret key file exists, it is read from the file. Otherwise, a new key is generated and
+/// written to the file.
+fn init_authenticator(
+    init_seed: [u8; 32],
+    secret_key_path: impl AsRef<Path>,
+) -> Result<(AuthScheme, BasicAuthenticator<StdRng>), ClientError> {
+    // Load secret key from file or generate new one
+    let secret = if secret_key_path.as_ref().exists() {
+        SecretKey::read_from_bytes(
+            &std::fs::read(secret_key_path).context("Failed to read secret key from file")?,
+        )
+        .map_err(ImplError)
+        .context("Failed to deserialize secret key")?
+    } else {
+        let mut rng = ChaCha20Rng::from_seed(init_seed);
+        let secret = SecretKey::with_rng(&mut rng);
+        std::fs::write(secret_key_path, secret.to_bytes())
+            .context("Failed to write secret key to file")?;
+
+        secret
+    };
 
-    /// Updates the stored faucet account with the provided delta.
-    fn update_faucet_account(&mut self, delta: &AccountDelta) -> Result<(), ProcessError> {
-        self.faucet_account
-            .borrow_mut()
-            .apply_delta(delta)
-            .map_err(|err| ProcessError::InternalServerError(err.to_string()))
-    }
-}
+    let auth_scheme = AuthScheme::RpoFalcon512 { pub_key: secret.public_key() };
 
-impl DataStore for FaucetDataStore {
-    fn get_transaction_inputs(
-        &self,
-        account_id: AccountId,
-        _block_ref: u32,
-        _notes: &[NoteId],
-    ) -> Result<TransactionInputs, DataStoreError> {
-        let account = self.faucet_account.borrow();
-        if account_id != account.id() {
-            return Err(DataStoreError::AccountNotFound(account_id));
-        }
-
-        let empty_input_notes =
-            InputNotes::new(Vec::new()).map_err(DataStoreError::InvalidTransactionInput)?;
-
-        TransactionInputs::new(
-            account.clone(),
-            account.is_new().then_some(self.seed),
-            self.block_header,
-            self.chain_mmr.clone(),
-            empty_input_notes,
-        )
-        .map_err(DataStoreError::InvalidTransactionInput)
-    }
-}
+    let authenticator = BasicAuthenticator::<StdRng>::new(&[(
+        secret.public_key().into(),
+        AuthSecretKey::RpoFalcon512(secret),
+    )]);
 
-// HELPER FUNCTIONS
-// ================================================================================================
+    Ok((auth_scheme, authenticator))
+}
 
 /// Builds a new faucet account with the provided configuration.
 ///
 /// Returns the created account, its seed, and the secret key used to sign transactions.
-fn build_account(config: FaucetConfig) -> Result<(Account, Word, SecretKey), InitError> {
+fn build_account(
+    config: &FaucetConfig,
+    init_seed: [u8; 32],
```

Comment:
Good catch, thank you! This is not the only issue here I can see so far. One of the most important things here is a vulnerability with a secret key generation. Since we initialize random number generator with always the same seed, we get the same secret for each faucet. This means, everybody can generate it by themselves and interact with the faucet. We should fix it before merging of this PR.

Reply:
`SecretKey` generation yields the public key which is stored on the account's storage which means it will affect the storage commitment, and thus the account seed (3 in your list) as well.